server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java

```diff
@@ -0,0 +1,176 @@
+                threadContext.markAsSystemContext();
+                awaitDiscovery();
+            }
+        }
+    }
+
+    public void stop() {
+        assert running == true;
+        running = false;
+    }
+
+    private void awaitDiscovery() {
+        if (running == false) {
+            logger.debug("awaitDiscovery: not running");
+            return;
+        }
+
+        final GetDiscoveredNodesRequest request = new GetDiscoveredNodesRequest();
+        request.setWaitForNodes(initialMasterNodeCount);
+        logger.trace("sending {}", request);
+        transportService.sendRequest(transportService.getLocalNode(), GetDiscoveredNodesAction.NAME, request,
+            new TransportResponseHandler<GetDiscoveredNodesResponse>() {
+                @Override
+                public void handleResponse(GetDiscoveredNodesResponse response) {
+                    assert response.getNodes().size() >= initialMasterNodeCount;
+                    assert response.getNodes().stream().allMatch(DiscoveryNode::isMasterNode);
+
+                    logger.debug("discovered {}, starting to bootstrap", response.getNodes());
+                    awaitBootstrap(response.getBootstrapConfiguration());
+                }
+
+                @Override
+                public void handleException(TransportException exp) {
+                    if (exp.getRootCause() instanceof ElasticsearchTimeoutException) {
+                        logger.debug(new ParameterizedMessage("discovery attempt timed out, retrying, request={}", request), exp);
+                        awaitDiscovery();
+                    } else {
+                        // exceptions other than a timeout are fatal
+                        logger.warn("discovery attempt failed, not retrying", exp);
+                    }
+                }
+
+                @Override
+                public String executor() {
+                    return Names.SAME;
+                }
+
+                @Override
+                public GetDiscoveredNodesResponse read(StreamInput in) throws IOException {
+                    return new GetDiscoveredNodesResponse(in);
+                }
+            });
+    }
+
+    private void awaitBootstrap(final BootstrapConfiguration bootstrapConfiguration) {
+        if (running == false) {
+            logger.debug("awaitBootstrap: not running");
+            return;
+        }
+
+        BootstrapClusterRequest request = new BootstrapClusterRequest(bootstrapConfiguration);
+        logger.trace("sending {}", request);
+        transportService.sendRequest(transportService.getLocalNode(), BootstrapClusterAction.NAME, request,
+            new TransportResponseHandler<BootstrapClusterResponse>() {
+                @Override
+                public void handleResponse(BootstrapClusterResponse response) {
+                    logger.debug("bootstrapped successful: received {}", response);
```

Comment:
successfully?

Reply:
Fixed the message in 12d90d2600a - I meant "bootstrapping" not "bootstrapped".